src/rust/engine/graph/src/entry.rs

```diff
@@ -108,20 +110,38 @@ pub(crate) enum EntryState<N: Node> {
   // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
   // order to compute the generation value for this Node by comparing it to the new result the next
   // time the Node runs.
-  NotStarted {
+  //
+  // Similarly, because exceptions are not cached in the graph, previous_error does not affect the
+  // value for this Entry, and is only used in .peek() so we can e.g. generate a scheduler stack
+  // trace before exiting from an uncaught exception. If the exception is caught, this value is
+  // ignored and eventually cleared when this Entry completes successfully.
+  Dormant {
     run_token: RunToken,
     generation: Generation,
     previous_result: Option<EntryResult<N>>,
+    previous_error: Option<N::Error>,
+  },
+  // A node which previously received an error which indicated that the node would be unable to
+  // satisfy any further inputs. This will occur if a node ever errors out with a cyclic graph
+  // exception.
+  //
+  // This exists as a distinct EntryState so that we can accurately reconstruct the graph to
```

Comment:
Hm. I'm not sure what "exit" means here... in particular, `pantsd` doesn't exit, so these will continue to exist in the graph, which I think means that there isn't a way to clear them...?

I think that you'd probably want to adjust the definition of "is_dirty"/"is_clean" similar to how #9015 did: https://github.com/pantsbuild/pants/pull/9015/files#diff-75bfebc7637079e53ddae93c399f8495L71-R84 ... that way an Erroneous node would try running again on the next attempt.

I sortof think that it might be a good idea to land #9015 to get the `Session` stuff. I have a strong suspicion that it will make the behavior of the "previous_error" and "Erroneous" aspects of this easier. It will be a definite pain to resolve conflicts on though =/

Reply:
> Hm. I'm not sure what "exit" means here... in particular, pantsd doesn't exit,

In this case, I meant "exit" as in "I would expect there to be an uncaught exception from the cyclic graph error". 

> so these will continue to exist in the graph, which I think means that there isn't a way to clear them...?

Correct!